File: leetCode/queue/Number_of_Islands_200/Main.java
import java.util.Queue;

class Solution {

    private int[][] moves = {{0,1}, {0,-1}, {1,0}, {-1,0}};

    private boolean isInside(int i, int j, int m, int n) {
        return !(i >= m || j >= n || j < 0 || i < 0);
    }

    private void bfs(int i, int j, char[][] grid, boolean[][] visited) {
        Queue<int[]> cellQueue = new LinkedList<>();
        cellQueue.offer(new int[]{i, j});

        while(!cellQueue.isEmpty()) {
            int[] c = cellQueue.poll();
            visited[c[0]][c[1]] = true;
            for (int[] move : moves) {
                int x = c[0] + move[0];
                int y = c[1] + move[1];
                if(isInside(x, y, grid.length, grid[0].length) && 
                    grid[x][y] == '1' &&
                    !visited[x][y]) {
                    cellQueue.offer(new int[]{x, y});
                    visited[x][y] = true;
                }
            }
        }
    }

    private void dfs(int i, int j, char[][] grid, boolean[][] visited) {
        if(i >= grid.length || j >= grid[0].length || j < 0 || i < 0) {
            return;
        }
        if(visited[i][j]) return;
        if(grid[i][j] == '0') {
            visited[i][j] = true;
            return;
        }
        visited[i][j] = true;
        dfs(i - 1, j, grid, visited);
        dfs(i + 1, j, grid, visited);
        dfs(i, j + 1, grid, visited);
        dfs(i, j - 1, grid, visited);
    }

    public int numIslands(char[][] grid) {
        if (grid.length == 0) return 0;
        boolean visited[][] = new boolean[grid.length][grid[0].length];
        int numIslands = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                if(!visited[i][j] && grid[i][j] == '1') {
                    // start BFS
                    bfs(i, j, grid, visited);
                    numIslands++;
                    // increment a counter
                }
            }
        }
        return numIslands;
    }
}

public class Main {
    public static void main(String[] args) {
        
    }
}

File: leetCode/queue/Number_of_Islands_200/main.py
class Solution:
    def bfs(self, i, j, grid):
        q = []
        q.append((i, j))
        while q:
            x, y = q.pop(0)
            for move in self.moves:
                next_x = x + move[0]
                next_y = y + move[1]
                if self.is_inside(next_x, next_y) and grid[next_x][next_y] == '1':
                    q.append((next_x, next_y))
                    grid[next_x][next_y] = '0'

    def dfs(self, i, j, grid):
        if i < 0 or j < 0 or i >= len(grid) or j >= len(grid[0]):
            return
        if grid[i][j] == '0':
            return
        grid[i][j] = '0'
        self.dfs(i + 1, j, grid)
        self.dfs(i - 1, j, grid)
        self.dfs(i, j + 1, grid)
        self.dfs(i, j - 1, grid)

    def numIslands(self, grid: List[List[str]]) -> int:
        self.moves = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        self.is_inside = lambda i, j: not(i < 0 or j < 0 or i >= len(grid) or j >= len(grid[0]))

        res = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    self.bfs(i, j, grid)
                    res += 1
        return res